services/gateway/endpoints/src/main/java/org/eclipse/ditto/services/gateway/endpoints/routes/cloudevents/CloudEventsRoute.java

```diff
@@ -0,0 +1,281 @@
+                                    : CommandNotSupportedException.newBuilder(schemaVersion.toInt())
+                                    .dittoHeaders(dittoHeaders)
+                                    .build();
+                        } catch (final Exception e) {
+                            return new Status.Failure(e);
+                        }
+                    })
+                    .to(Sink.actorRef(createHttpPerRequestActor(ctx, httpResponseFuture),
+                            AbstractHttpRequestActor.COMPLETE_MESSAGE))
+            );
+
+            // return with future
+
+            return completeWithFuture(httpResponseFuture);
+
+        });
+
+    }
+
+    /**
+     * Convert the request and payload to a cloud event.
+     *
+     * @param ctx The request context.
+     * @param payload The binary payload, this may contain the cloud event payload, or the fully encoded cloud
+     *         event structure.
+     * @return The cloud event
+     */
+    private @Nullable
+    CloudEvent toCloudEvent(final RequestContext ctx, final DittoHeaders dittoHeaders, final ByteString payload) {
+
+        if (LOGGER.isTraceEnabled()) {
+            LOGGER.trace("Headers:");
+            ctx.getRequest().getHeaders().forEach(header -> LOGGER.trace("\t{} = {}", header.name(), header.value()));
+            LOGGER.trace("Ditto:");
+            dittoHeaders.forEach((k, v) -> LOGGER.trace("\t{} = {}", k, v));
+        }
+
+        // create a reader for the message
+        final MessageReader reader = HttpMessageFactory.createReader(acceptor -> {
+
+            // NOTE: this acceptor may be run multiple times by the message reader
+
+            // record if we saw the content type header
+            final AtomicBoolean sawContentType = new AtomicBoolean();
+            // consume the HTTP request headers
+            ctx.getRequest().getHeaders().forEach(header -> {
+                if (header.lowercaseName().equals(DittoHeaderDefinition.CONTENT_TYPE.getKey())) {
+                    sawContentType.set(true);
+                }
+                acceptor.accept(header.name(), header.value());
+            });
+
+            if (!sawContentType.get()) {
+                // we didn't see the content type in the header, so extract it from akka's request
+                acceptor.accept(DittoHeaderDefinition.CONTENT_TYPE.getKey(), ctx.getRequest().entity().getContentType().mediaType().toString());
+            }
+        }, payload.toArray());
+
+        return reader.toEvent();
+    }
+
+    private CloudEvent validateCloudEvent(@Nullable final CloudEvent cloudEvent, final RequestContext ctx, final DittoHeaders dittoHeaders) {
+
+        if (cloudEvent == null || cloudEvent.getData() == null) {
+            throw MissingPayloadException
+                    .withDetailedInformationBuilder()
+                    .dittoHeaders(dittoHeaders)
+                    .build();
+        }
+
+        LOGGER.debug("Cloud event: {}", cloudEvent);
```

Comment:
Please make use of `LOGGER.withCorrelationId(dittoHeaders).trace(...)` in order to log the correlation-id of the DittoHeaders to the MDC.